### ISSUE TITLE
Pin version

### ISSUE THREAD

**Issue by mstenta** [geerlingguy/ansible-role-docker#85]
Is there a way to pin to a specific version so that `apt-get upgrade` does not upgrade Docker and Docker Compose?

**Comment by geerlingguy** [comment]
@mstenta yes, if using apt/Debian/Ubuntu, you can use `packagename=version` for the docker package, for example, see another project: https://github.com/geerlingguy/raspberry-pi-dramble/blob/kubernetes/roles/docker-arm/tasks/main.yml#L19-L23

**Comment by geerlingguy** [comment]
So for this role you'd set something like:

```yaml
docker_version: 18.04.0~ce~3-0~raspbian
docker_package: "docker-{{ docker_edition }}={{ docker_version }}"
```

(Note that the raspbian version is for ARM... you need to look in your system's apt repo for valid version strings for your platform's flavors of Docker).

**Comment by mstenta** [comment]
Thanks @geerlingguy !

I have something like that, but it seems that `apt-get upgrade` is still trying to upgrade my `docker-ce` package.

Although... I had previously been managing the server with Puppet, and only recently transitioned to Ansible - so I will investigate to see if there's something left over that's causing this...

**Comment by mstenta** [comment]
Wait - sorry - scratch that... one of the files I removed was `/etc/apt/sources.list.d/download_docker_com_linux_ubuntu.list`, and when I run Ansible again that gets recreated...

When I run `apt-get upgrade`, `docker-ce` is shown as one of the packages that will be upgraded again.

Is there code in this repo that pins the package version so that it doesn't get upgraded by `apt-get upgrade`? Or is that something that should be done separately from the Docker installation process? Or do you think I'm doing something else wrong?

This is what I have for my role variables:

```
# Pin the Docker and Docker Compose versions.
# See https://github.com/geerlingguy/ansible-role-docker/issues/45
docker_version: 17.03.0
docker_package: docker-{{docker_edition}}={{docker_version}}~{{docker_edition}}-0~{{ansible_distribution|lower}}-{{ansible_distribution_release}}
docker_compose_version: 1.11.0
```

Here is information about my currently installed versions:

```
$ sudo docker --version
Docker version 17.03.0-ce, build 3a232c8
$ sudo docker-compose --version
docker-compose version 1.11.0, build 6de1806
```

And here is part of the output of `apt-get upgrade -s` (simulation):

```
The following packages will be upgraded:
  docker-ce
1 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
Inst docker-ce [17.03.0~ce-0~ubuntu-xenial] (18.06.0~ce~3-0~ubuntu Docker CE:xenial [amd64])
Conf docker-ce (18.06.0~ce~3-0~ubuntu Docker CE:xenial [amd64])
```

It looks like it wants to upgrade to 18.06.0...

**Comment by mstenta** [comment]
I solved this by adding another task to my Ansible playbook specifically for pinning the `docker-ce` version. This also required me to add an additional variable so that I had the version in the format that apt pinning expects.

Here are the variables I'm using:

```
docker_version: 17.03.0
docker_package_version: "{{docker_version}}~{{docker_edition}}-0~{{ansible_distribution|lower}}-{{ansible_distribution_release}}"
docker_package: "docker-{{docker_edition}}={{docker_package_version}}"
```

Notice that previously there was no `docker_package_version` variable (it was part of `docker_package`). But apt pinning needs the variable to just include the version, not the `docker-{{docker_edition}}=` part.

Here is the new task I added to pin the version:

```
- name: Pin the Docker version so that apt-get upgrade does not upgrade it.
  copy:
    content: "Package: docker-ce\nPin: version {{ docker_package_version }}\nPin-Priority: 1000\n"
    dest: /etc/apt/preferences.d/docker
    owner: root
    group: root
    mode: 0644
```

This does not currently handle pinning of docker-compose.

@geerlingguy Should we reopen this issue? As far as I can tell, the Docker role does not actually pin the version. It let's you specify which version to install, but doesn't prevent it from being upgraded by `apt-get upgrade`.
